### PR TITLE
Print logic axes

### PR DESCRIPTION
**Pull request by Shuang-cnt**
# Description

This change builds upon [PR#2866](https://github.com/AI-Hypercomputer/maxtext/pull/2866) and [PR#2926](https://github.com/AI-Hypercomputer/maxtext/pull/2926) to add functionality for printing logical axes

# Tests

Command:
` python -m MaxText.train_compile MaxText/configs/base.yml compile_topology=v5p-1024 compile_topology_num_slices=1 model_name=deepseek3-671b per_device_batch_size=1  ici_tensor_parallelism=8 ici_expert_parallelism=8  log_config=false debug_sharding=true`

## Output example:

Log for a real train: https://paste.googleplex.com/6519526529302528

### Weight with Logic axes:
```
I0123 00:00:29.309755 133441853451392 maxtext_utils.py:1269]  params/params/decoder/dense_layers/mlp/wi_0/kernel
    Shape:     float32[7168,3,18432]
    Logical:   PartitionSpec('embed', 'dense_layers', 'mlp')
    Physical:  (('fsdp', 'expert'), None, 'tensor')
I0123 00:00:29.309807 133441853451392 maxtext_utils.py:1269]  params/params/decoder/dense_layers/mlp/wi_1/kernel
    Shape:     float32[7168,3,18432]
    Logical:   PartitionSpec('embed', 'dense_layers', 'mlp')
    Physical:  (('fsdp', 'expert'), None, 'tensor')
I0123 00:00:29.309876 133441853451392 maxtext_utils.py:1269]  params/params/decoder/dense_layers/mlp/wo/kernel
    Shape:     float32[18432,3,7168]
    Logical:   PartitionSpec('mlp', 'dense_layers', 'embed')
    Physical:  ('tensor', None, ('fsdp', 'expert'))

```
### Activation Logical Axes:
```
I0126 22:03:30.305064 128637779627136 deepseek.py:131] Logical:  bfloat16[512,2048,7168]..................................... ('activation_batch', 'activation_norm_length', 'activation_embed')
I0126 22:03:30.305159 128637779627136 deepseek.py:131] bfloat16[512,2048,7168]......................................................... (('fsdp', 'expert'), None, 'tensor').
I0126 22:03:30.309315 128637779627136 attention_mla.py:736] Logical:  bfloat16[512,2048,7168]..................................... ('activation_batch', 'activation_length_no_exp', 'activation_embed')
I0126 22:03:30.309363 128637779627136 attention_mla.py:736] bfloat16[512,2048,7168]......................................................... (('fsdp', 'expert'), None, 'tensor').
I0126 22:03:30.312696 128637779627136 attention_mla.py:538] Logical:  bfloat16[512,2048,128,128].................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim')
I0126 22:03:30.312744 128637779627136 attention_mla.py:538] bfloat16[512,2048,128,128]...................................................... (('fsdp', 'expert'), None, 'tensor', None).
```
# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/maxtext_utils.py [hunk 1]
```diff
@@ -1204,12 +1207,101 @@ def schedule(step):
...
+    return
+
+  # Build a reverse map (Potential Physical Axes Tuple -> List of Semantic Names)
+  rule_value_to_semantic = defaultdict(list)
+  if logical_axis_rules:
+    rules_iter = logical_axis_rules.items() if isinstance(logical_axis_rules, dict) else logical_axis_rules
+    for name, potentials in rules_iter:
+      # name: LHS for example 'embed/activation_batch
+      # potentials: RHS for example 'data', 'model', None, ['data', 'model']
+      if isinstance(potentials, str):
```
Comment (richjames0) [inline src/MaxText/maxtext_utils.py]: nit: match statement is probably cleaner here

File: src/MaxText/maxtext_utils.py [hunk 2]
```diff
@@ -1204,12 +1207,99 @@ def schedule(step):
...
+        semantic_str = "Partitionspec(" + ", ".join(semantic_parts) + ")"
+      elif leaf_rule_value is None:  # Explicit None in rule
+        semantic_str = "Partitionspec(None)"
+      else:
+        semantic_str = str(leaf_rule_value)
+    else:
+      semantic_str = "None (No Logical Info)"
+
+    # Multi-line logging
+    max_logging.info(f"{path_str}")
```
Comment (xibinliu) [inline src/MaxText/maxtext_utils.py]: Use one log entry for one path. Header in line 1246 can be removed.
```
    message = (
        f" {path_str} sharding info:\n"
        f"    Shape:     {shape}\n"
        f"    Logical:   {semantic_str}\n"
        f"    Physical:  {pspec_str}"
    )
    max_logging.info(message)
```

File: src/MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -1204,12 +1207,99 @@ def schedule(step):
...
+  else:
+    leaves_rule_values = [(None, None)] * len(leaves_params)
+
+  if not len(leaves_params) == len(leaves_sharding) == len(leaves_rule_values):
+    max_logging.warning(
+        "Warning: Parameter tree structure mismatch between params, shardings," " and logical annotations."
+    )
+    return
+
+  # Build a reverse map
```
Comment (xibinliu) [inline src/MaxText/maxtext_utils.py]: state_logical_annotations should have the only logic rule for a given path, so there is no need to do reverse map.

File: src/MaxText/sharding.py [hunk 4]
```diff
@@ -59,18 +61,118 @@ def maybe_shard_with_name(inputs, named_sharding, shard_mode, debug_sharding=Fal
...
     return None
-  named_sharding = create_sharding(mesh, logical_axes, rules=rules)
-  return maybe_shard_with_name(
-      inputs, named_sharding, shard_mode, debug_sharding=debug_sharding, extra_stack_level=extra_stack_level + 1
-  )
+
+  active_rules = rules if rules is not None else _GLOBAL_LOGICAL_RULES
+  named_sharding = create_sharding(mesh, logical_axes, rules=active_rules)
+
+  if debug_sharding and hasattr(inputs, "shape"):
```
Comment (xibinliu) [inline src/MaxText/sharding.py]: Can we move the logging into the maybe_shard_with_name and modify the existing activation sharding log to include logical sharding info?

File: src/MaxText/sharding.py [hunk 5]
```diff
@@ -59,18 +61,118 @@ def maybe_shard_with_name(inputs, named_sharding, shard_mode, debug_sharding=Fal
...
+    max_logging.info("=" * 120)
+
+    stack = inspect.stack()
+    caller_info = _get_caller_info(stack, 2 + extra_stack_level)
+    unresolved_str = _format_rule_info(active_rules, logical_axes)
+    resolved_str = _format_phys_spec_info(named_sharding, logical_axes)
+    shape_str = str(jax.typeof(inputs))
+
+    # format: Caller [RULES] -> [FINAL] Shape
+    max_logging.info(f"{caller_info}")
```
Comment (xibinliu) [inline src/MaxText/sharding.py]: We do need something to identify which module is this activation for. No sure if this can be identified with the existing stack_level parameter.

File: src/MaxText/sharding.py [hunk 6]
```diff
@@ -59,18 +61,118 @@ def maybe_shard_with_name(inputs, named_sharding, shard_mode, debug_sharding=Fal
...
+  active_rules = rules if rules is not None else _GLOBAL_LOGICAL_RULES
+  named_sharding = create_sharding(mesh, logical_axes, rules=active_rules)
+
+  if debug_sharding and hasattr(inputs, "shape"):
+    max_logging.info("=" * 120)
+    max_logging.info("    Tracing logical axes for activations during JIT compilation.")
+    max_logging.info("=" * 120)
+
+    stack = inspect.stack()
+    caller_info = _get_caller_info(stack, 2 + extra_stack_level)
```
Comment (xibinliu) [inline src/MaxText/sharding.py]: no need to do the reverse mapping

File: src/MaxText/maxtext_utils.py (hunk 2 above)
Comment (Shuang-cnt) [reply src/MaxText/maxtext_utils.py]: Done. Thanks！

File: src/MaxText/maxtext_utils.py (hunk 3 above)
Comment (Shuang-cnt) [reply src/MaxText/maxtext_utils.py]: Done. Thanks！

File: src/MaxText/sharding.py (hunk 4 above)
Comment (Shuang-cnt) [reply src/MaxText/sharding.py]: Done. Thanks！

File: src/MaxText/sharding.py (hunk 5 above)
Comment (Shuang-cnt) [reply src/MaxText/sharding.py]: check with Nuojin, currently , they expect to see  file names with line numbers for example: `attention_mla.py:544]` in the log

File: src/MaxText/sharding.py (hunk 6 above)
Comment (Shuang-cnt) [reply src/MaxText/sharding.py]: Done, thanks!

File: src/MaxText/maxtext_utils.py (hunk 1 above)
Comment (Shuang-cnt) [reply src/MaxText/maxtext_utils.py]: Updated the maxtext_utils.py. Done. Thanks

File: src/MaxText/max_utils.py [hunk 7]
```diff
@@ -1032,3 +1032,14 @@ def transformer_engine_context():
...
+
+def print_mesh_axes_info(mesh: jax.sharding.Mesh):
+  """Prints all mesh axes and their sizes in a single comma-separated line."""
+  if not mesh.shape:
+    max_logging.info("Mesh Axes: (Empty Mesh)")
+    return
+
+  axis_info = [f"{axis_name}: {axis_size}" for axis_name, axis_size in mesh.shape.items()]
+  info_str = "Mesh Axes: (" + ", ".join(axis_info) + ")"
+  max_logging.info(info_str)
```
Comment (NuojCheng) [inline src/MaxText/max_utils.py]: This function seems replicated with `print_non_trivial_mesh_axis`. I personally would prefer only printing mesh axes with size larger than one.

File: src/MaxText/maxtext_utils.py [hunk 8]
```diff
@@ -944,6 +944,16 @@ def setup_initial_state(
   return state, state_mesh_annotations, state_mesh_shardings, data_iterator
 
 
+def get_logical_annotations(model, tx, config, rng, mesh, is_training=True):
+  init_state_partial = functools.partial(init_initial_state, model, tx, config, is_training, rng)
+
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
```
Comment (NuojCheng) [inline src/MaxText/maxtext_utils.py]: use `with jax.set_mesh(mesh), nn_partitioning.axis_rules(config.logical_axis_rules):`

File: src/MaxText/maxtext_utils.py [hunk 9]
```diff
@@ -1227,15 +1237,33 @@ def schedule(step):
   return optax.join_schedules(pieces, boundaries)
 
 
-def print_shardings_params(params, params_sharding, mesh):
-  """Print state shardings."""
+def print_shardings_params(params, params_sharding, mesh, logical_annotations=None):
+  """
+  Print state shardings comparing Logical Definition vs Physical Result.
+  Simplified version: Directly prints logical annotations without reverse mapping.
```
Comment (NuojCheng) [inline src/MaxText/maxtext_utils.py]: what does simplified version mean here?

Comment (Shuang-cnt) [reply src/MaxText/maxtext_utils.py]: forgot to delete it.  I will delete it, thanks.

File: src/MaxText/sharding.py
Comment (NuojCheng) [inline src/MaxText/sharding.py]: Could you add `Physical: ` here for consistency?

File: src/MaxText/sharding.py [hunk 10]
```diff
@@ -67,9 +68,26 @@ def maybe_shard_with_logical(
...
+      max_logging.info(f"Logical:  {log_key[0]:.<60} {log_key[1]}", stacklevel=3 + extra_stack_level)
+      max_logging.info(f"{log_key[0]:.<80} {pspec_str}.", stacklevel=3 + extra_stack_level)
+      _LOGGED_LOGICAL_AXES.add(log_key)
+
   return maybe_shard_with_name(
-      inputs, named_sharding, shard_mode, debug_sharding=debug_sharding, extra_stack_level=extra_stack_level + 1
+      inputs,
+      named_sharding,
+      shard_mode,
+      debug_sharding=False,
```
Comment (NuojCheng) [inline src/MaxText/sharding.py]: `debug_sharding=debug_sharding`.

We still want to print physical partitions even when logical partitions are printed.

File: src/MaxText/max_utils.py (hunk 7 above)
Comment (Shuang-cnt) [reply src/MaxText/max_utils.py]: Done

File: src/MaxText/maxtext_utils.py (hunk 8 above)
Comment (Shuang-cnt) [reply src/MaxText/maxtext_utils.py]: Done

File: src/MaxText/sharding.py
Comment (Shuang-cnt) [reply src/MaxText/sharding.py]: Done

File: src/MaxText/sharding.py (hunk 10 above)
Comment (Shuang-cnt) [reply src/MaxText/sharding.py]: Done

**Comment by Shuang-cnt** [comment]
> Thank you for the PR! Made some comments and suggestions. It would be nice to add a real train test in the PR description.

Thanks for the suggestion. Add a real train log. Please check.
